### DISCUSSION TITLE
How to use a certificate to connect to an ES cluster

### DISCUSSION THREAD

**Discussion by ktpktr0** [jertel/elastalert2#586] [Q&A]
I want to know how elasalert uses certificates to connect to es. I tried the following methods, but it doesn't work.

To generate a certificate, I use the existing p12 certificate (This certificate is used to generate the node certificate of the ES cluster)
```
echo "" | /usr/share/elasticsearch/bin/elasticsearch-certutil cert --silent --pem /etc/elasticsearch/ssl/elastic-stack-ca.p12 --ip 192.168.1.165 --dns es-master2 --out /tmp/1.zip
```
elastalert.yaml:
```
#Optional, select whether to connect es with SSL, true or false
use_ ssl: True

#Optional, whether to verify the TLS certificate. Set it to true or false. The default is - true
#verify_ certs: False

#Select certificate
ca_ certs: /home/elastalert/ssl/ca.crt
client_ crt: /home/elastalert/ssl/es-master2.crt
client_ key: /home/elastalert/ssl/es-master2.key
```
error log:
```
Traceback (most recent call last):
  File "/usr/local/bin/elastalert-create-index", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.10/site-packages/elastalert/create_index.py", line 278, in main
    create_index_mappings(es_client=es, ea_index=index, recreate=args.recreate, old_ea_index=old_index)
  File "/usr/local/lib/python3.10/site-packages/elastalert/create_index.py", line 23, in create_index_mappings
    esinfo = es_client.info()['version']
  File "/usr/local/lib/python3.10/site-packages/elasticsearch/client/utils.py", line 84, in _wrapped
    return func(*args, params=params, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/elasticsearch/client/__init__.py", line 259, in info
    return self.transport.perform_request("GET", "/", params=params)
  File "/usr/local/lib/python3.10/site-packages/elasticsearch/transport.py", line 318, in perform_request
    status, headers_response, data = connection.perform_request(method, url, params, body, headers=headers, ignore=ignore, timeout=timeout)
  File "/usr/local/lib/python3.10/site-packages/elasticsearch/connection/http_requests.py", line 83, in perform_request
    raise SSLError('N/A', str(e), e)
elasticsearch.exceptions.SSLError: ConnectionError(HTTPSConnectionPool(host='192.168.1.165', port=9200): Max retries exceeded with url: / (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self signed certificate in certificate chain (_ssl.c:997)')))) caused by: SSLError(HTTPSConnectionPool(host='192.168.1.165', port=9200): Max retries exceeded with url: / (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self signed certificate in certificate chain (_ssl.c:997)'))))
``
```
My es cluster node uses the following command to generate an xpack certificate:

echo "" | /usr/share/elasticsearch/bin/elasticsearch-certutil cert --ca elastic-stack-ca.p12 --ip 192.168.1.165 --dns es-master2 --out es-master2.p12 --pass ""
```

However, generating the certificate as at the beginning does not seem to work properly. I don't know how elasalert needs to generate a certificate

**Comment by ktpktr0** [comment]
I've tried, but it didn't work

[johnsusek/praeco#345 (comment)](https://github.com/johnsusek/praeco/issues/345#issuecomment-822827568)

**Reply by nsano-rururu** [reply]
@ktpktr0 

Please tell me what part of the link you executed.

**Reply by nsano-rururu** [reply]
https://kazuhira-r.hatenablog.com/entry/2020/02/16/015432

**Reply by nsano-rururu** [reply]
Secure the Elastic Stack
https://www.elastic.co/guide/en/elasticsearch/reference/current/secure-cluster.html

**Reply by ktpktr0** [reply]
> Secure the Elastic Stack https://www.elastic.co/guide/en/elasticsearch/reference/current/secure-cluster.html

My es cluster works well with certificates. I don't need to care about the certificates of the ES cluster. What I want to know now is what kind of certificate elastalrt needs and how it is generated. Looking at my previous example, I want to know why it is wrong.

**Reply by nsano-rururu** [reply]
Starting with Python 2.7.9, if the SSL certificate is incorrect, it will give an error by default. The error message is an error that the SSL certificate is incorrect.

PEP 476 -- Enabling certificate verification by default for stdlib http clients
https://www.python.org/dev/peps/pep-0476/

**Reply by ktpktr0** [reply]
> Starting with Python 2.7.9, if the SSL certificate is incorrect, it will give an error by default. The error message is an error that the SSL certificate is incorrect.
> 
> PEP 476 -- Enabling certificate verification by default for stdlib http clients https://www.python.org/dev/peps/pep-0476/

It cannot be solved for the time being

**Comment by ktpktr0** [comment]
Through research, I have solved the problem of using SSL to connect the ES cluster with elasalert. The following is the method to generate the certificate:

https://github.com/Yelp/elastalert/issues/3228

**Reply by nsano-rururu** [reply]
I couldn't understand why I wrote comments here and there.
